tests/test_transaction.py

```diff
@@ -246,39 +286,172 @@ def test_nested_inner_scope_exception_handled_in_outer_scope(conn, svcconn):
+        insert_row(conn, "foo")
+    assert_rows(conn, set())
+    assert_rows(svcconn, set())
 
 
-def assert_not_in_transaction(conn):
-    assert conn.pgconn.transaction_status == conn.TransactionStatus.IDLE
+def test_force_rollback_exception_exit(conn, svcconn):
+    """
+    Transaction started with the force_rollback option enabled discards all
+    changes at the end of the context.
+    """
+    with pytest.raises(ExpectedException):
+        with conn.transaction(force_rollback=True):
+            insert_row(conn, "foo")
+            raise ExpectedException()
+    assert_rows(conn, set())
+    assert_rows(svcconn, set())
 
 
-def assert_in_transaction(conn):
-    assert conn.pgconn.transaction_status == conn.TransactionStatus.INTRANS
+def test_explicit_rollback_discards_changes(conn, svcconn):
+    """
+    Raising a Rollback exception in the middle of a block exits the block and
+    discards all changes made within that block.
+
+    You can raise any of the following:
+     - Rollback (type)
+     - Rollback() (instance)
+     - Rollback(tx) (instance initialised with reference to the transaction)
+     - tx.rollback_exception (convenience property that returns Rollback(tx))
+    All of these are equivalent.
+    """
+    tx = conn.transaction()
+    for to_raise in (
+        Rollback,
+        Rollback(),
+        Rollback(tx),
+        tx.rollback_exception,
+    ):
+        with tx:
+            insert_row(conn, "foo")
+            raise to_raise
+        assert_rows(conn, set(""))
+        assert_rows(svcconn, set())
 
 
-class ExpectedException(Exception):
-    pass
+def test_explicit_rollback_outer_transaction_unaffected(conn, svcconn):
+    """
+    Raising a Rollback exception in the middle of a block does not impact an
+    enclosing transaction block.
+    """
+    with conn.transaction():
+        insert_row(conn, "before")
+        with conn.transaction():
+            insert_row(conn, "during")
+            raise Rollback
+        assert_in_transaction(conn)
+        assert_rows(svcconn, set())
+        insert_row(conn, "after")
+    assert_rows(conn, {"before", "after"})
+    assert_rows(svcconn, {"before", "after"})
+
+
+def test_explicit_rollback_of_outer_transaction(conn):
+    """
+    Raising a Rollback exception that references an outer transaction will
+    discard all changes from both inner and outer transaction blocks.
+
+    This can be raised as:
+    - Rollback(outer_tx); or
+    - outer_tx.rollback_exception
+    Both of these are equivalent.
+    """
+    outer_tx = conn.transaction()
+    for to_raise in (Rollback(outer_tx), outer_tx.rollback_exception):
+        with outer_tx:
+            insert_row(conn, "outer")
+            with conn.transaction():
+                insert_row(conn, "inner")
+                raise to_raise
+            insert_row(conn, "unreachable")
+        assert_rows(conn, set())
+
+
+def test_explicit_rollback_of_enclosing_transaction_outer_transaction_unaffected(
```

Comment:
Agh, so long names that the arguments get pushed the line below? Too long imo: a docstring if you think a shorter name wouldn't explain enough would be better.